File: lista-de-exercicios/exercicio-13/Aluno.test.ts
import { jest } from '@jest/globals';
import { Aluno } from './Aluno';

describe('Classe Aluno', () => {
    test('A nota não pode ser menor que zero', () => {
        const alunoMenor = new Aluno(-10);

        expect(alunoMenor.getNotas()).toStrictEqual([0]);
    })

    test('A nota não pode ser maior que dez', () => {
        const alunoMaior = new Aluno(20);

        expect(alunoMaior.getNotas()).toStrictEqual([10]);
    })

    test('Deve adicionar uma nova nota ao grupo de notas', () => {
        const aluno = new Aluno(1);
        aluno.adicionarNota(5);

        expect(aluno.getNotas()).toStrictEqual([1, 5]);
    })

    test('Deve retornar a média das notas', () => {
        const alunoMedia = new Aluno(4);
        alunoMedia.adicionarNota(6);

        expect(alunoMedia.getMedia()).toBe(5);
    })

    test('Deve retornar o status correto do aluno', () => {
        const alunoAprovado = new Aluno(6.1);

        expect(alunoAprovado.getStatus()).toBe('Aprovado');

        const alunoSuplementar = new Aluno(6);
        const alunoSuplementar2 = new Aluno(4);

        expect(alunoSuplementar.getStatus()).toBe('Verificacao Suplementar');
        expect(alunoSuplementar2.getStatus()).toBe('Verificacao Suplementar');

        const alunoReprovado = new Aluno(3.9);

        expect(alunoReprovado.getStatus()).toBe('Reprovado');
    })

    test('Deve retornar o status correto após adicionar uma nova nota', () => {
        const aluno = new Aluno(6);

        expect(aluno.getStatus()).toBe('Verificacao Suplementar');

        aluno.adicionarNota(7);

        expect(aluno.getStatus()).toBe('Aprovado');
    })
})

File: lista-de-exercicios-java/src/test/java/org/example/exercicio13/AlunoTest.java
package org.example.exercicio13;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.ArrayList;

import static org.junit.jupiter.api.Assertions.*;

class AlunoTest {

    Aluno aluno = new Aluno();
    @DisplayName("A nota não pode ser menor que zero")
    @Test
    void notaDeveSerIgualOuMaiorQueZero() {
        assertThrows(IllegalArgumentException.class, () -> {
            aluno.adicionarNota(-10);
        });
    }

    @DisplayName("A nota não pode ser maior que dez")
    @Test
    void notaDeveSerIgualOuMaiorQueDez() {
        assertThrows(IllegalArgumentException.class, () -> {
            aluno.adicionarNota(20);
        });
    }

    @DisplayName("Deve adicionar uma nova nota ao grupo de notas")
    @Test
    void a() {
        aluno.adicionarNota(5);

        assertArrayEquals(new Double[] {5.0}, aluno.getNotas().toArray());
    }

    @DisplayName("Deve retornar a média das notas")
    @Test
    void deveRetornarMediaDeNotas() {
        aluno.adicionarNota(1);
        aluno.adicionarNota(5);

        assertEquals(3.0, aluno.calcularMedia());
    }

    @DisplayName("Deve retornar o status correto do aluno")
    @Test
    void deveRetornarStatusCorreto() {
        Aluno alunoAprovado = new Aluno();
        alunoAprovado.adicionarNota(6.1);
        Aluno alunoSuplementar = new Aluno();
        alunoSuplementar.adicionarNota(6);
        Aluno alunoSuplementar2 = new Aluno();
        alunoSuplementar2.adicionarNota(4);
        Aluno alunoReprovado = new Aluno();
        alunoReprovado.adicionarNota(3.9);

        assertEquals("Aprovado", alunoAprovado.calcularStatus());
        assertEquals("Verificação Suplementar", alunoSuplementar.calcularStatus());
        assertEquals("Verificação Suplementar", alunoSuplementar2.calcularStatus());
        assertEquals("Reprovado", alunoReprovado.calcularStatus());
    }

    @DisplayName("Deve retornar o status correto após adicionar uma nova nota")
    @Test
    void deveRetornarStatusCorretoAposAdicionarNota() {
        aluno.adicionarNota(1);
        aluno.adicionarNota(5);

        assertEquals("Reprovado", aluno.calcularStatus());
    }
}